<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [1]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat==4.2.0

print("done")


                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.2) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ['bs4==4.10.0']

[+] 0.0s
[+] 0.1s
pkgs/main/linux-64 ╸━━━━━━━━━━━━━━━╸━━━━━━━━   0.0 B /  ??.?MB @ 

In [42]:
# import packages

import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import numpy as np
import codecs

print("done")

done


## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [3]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [4]:
# extract stock data on ticker "TSLA" 

tesla = yf.Ticker("TSLA")


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [5]:
# put stock history for TSLA into variable tesla_data

tesla_data = tesla.history(period = "max")

tesla_data

Open        High         Low       Close     Volume  \
Date                                                                    
2010-06-29    1.266667    1.666667    1.169333    1.592667  281494500   
2010-06-30    1.719333    2.028000    1.553333    1.588667  257806500   
2010-07-01    1.666667    1.728000    1.351333    1.464000  123282000   
2010-07-02    1.533333    1.540000    1.247333    1.280000   77097000   
2010-07-06    1.333333    1.333333    1.055333    1.074000  103003500   
...                ...         ...         ...         ...        ...   
2023-10-25  215.880005  220.100006  212.199997  212.419998  107065100   
2023-10-26  211.320007  214.800003  204.880005  205.759995  115112600   
2023-10-27  210.600006  212.410004  205.770004  207.300003   94728100   
2023-10-30  209.279999  210.880005  194.669998  197.360001  136448200   
2023-10-31  196.119995  202.800003  194.070007  200.839996  116248372   

            Dividends  Stock Splits  
Date                                 
2010-06-29          0           0.0  
2010-06-30          0           0.0  
2010-07-01          0           0.0  
2010-07-02          0           0.0  
2010-07-06          0           0.0  
...               ...           ...  
2023-10-25          0           0.0  
2023-10-26          0           0.0  
2023-10-27          0           0.0  
2023-10-30          0           0.0  
2023-10-31          0           0.0  

[3359 rows x 7 columns]

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [6]:
# reset the index so that it is numbered

tesla_data.reset_index(inplace=True)

tesla_data

Date        Open        High         Low       Close     Volume  \
0    2010-06-29    1.266667    1.666667    1.169333    1.592667  281494500   
1    2010-06-30    1.719333    2.028000    1.553333    1.588667  257806500   
2    2010-07-01    1.666667    1.728000    1.351333    1.464000  123282000   
3    2010-07-02    1.533333    1.540000    1.247333    1.280000   77097000   
4    2010-07-06    1.333333    1.333333    1.055333    1.074000  103003500   
...         ...         ...         ...         ...         ...        ...   
3354 2023-10-25  215.880005  220.100006  212.199997  212.419998  107065100   
3355 2023-10-26  211.320007  214.800003  204.880005  205.759995  115112600   
3356 2023-10-27  210.600006  212.410004  205.770004  207.300003   94728100   
3357 2023-10-30  209.279999  210.880005  194.669998  197.360001  136448200   
3358 2023-10-31  196.119995  202.800003  194.070007  200.839996  116248372   

      Dividends  Stock Splits  
0             0           0.0  
1             0           0.0  
2             0           0.0  
3             0           0.0  
4             0           0.0  
...         ...           ...  
3354          0           0.0  
3355          0           0.0  
3356          0           0.0  
3357          0           0.0  
3358          0           0.0  

[3359 rows x 8 columns]

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [47]:
# get revenue data for TSLA from the given url

url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"

# save this as a text file

html_data = requests.get(url).text

#with open("out.txt", "w") as f:
    #f.write(html_data)
    
#html_data_tsla = codecs.open("out.txt", encoding = "utf-8")

#html_data_tsla

#html_data

Parse the html data using `beautiful_soup`.


In [8]:
#parse data using Beautiful Soup

soup = BeautifulSoup(html_data, "html.parser")

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [72]:
#create dataframe

tesla_revenue = pd.DataFrame(columns=["Date", "Revenue"])

#identify the tables in the data

tables = soup.find_all('table')

#since tables[1] was identified with having revenue data, we take this and identify the rows of data in this
table1 = tables[1].find_all('tr')  

#create an empty list to put the raw data into
data_list = []

# put the raw data into the list. This will contain alternating date and revenue data in a single list
for row in table1:
    row_data = row.find_all('td')
    individual_row_data = [data.text.strip() for data in row_data]
    data_list = data_list + individual_row_data

    
#dates_list: pick out the dates as every other entry in the data list starting with the very first one

dates_list = data_list[0:len(data_list):2]


#revenue_list: pick out the revenue as every other entry in the data list starting with the second one

revenue_list = data_list[1:len(data_list):2]



Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [77]:
#below code did not work for me, so I remove the $ sign a different way
#tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")

i = 0
revenue_list_nodolla = []
while(i<len(revenue_list)):
    revenue_list_nodolla.append(revenue_list[i].replace("$",""))
    i = i+1
    

revenue_list_nocomma = []

i = 0
while(i<len(revenue_list_nodolla)):
      revenue_list_nocomma.append(revenue_list_nodolla[i].replace(",",""))
      i = i+1
      
print(revenue_list_nocomma)

#revenue_list_float = []

#i = 0
#while(i<len(revenue_list_nocomma)):
    #newnumba = 0
    #newnumba = np.float64(revenue_list_nocomma[i])
    #print(newnumba)
    #revenue_list_int.append(newnumba)
    #i = i+1

revenue_list_nospace = []
i = 0
while(i<len(revenue_list_nocomma)):
    revenue_list_nospace.append(revenue_list_nocomma[i].replace(' ',''))
    i = i+1    

#print(revenue_list_nospace)
#print(revenue_list_nocomma)
        

    
# fill in data frame by creating a list for each row based on the existing lists for date and revenue

tesla_revenue_new = pd.DataFrame(columns = ["Date","Revenue"])
i = 0


while(i<len(dates_list)):
        new_row = [dates_list[i],float(revenue_list_nospace[i])]
        tesla_revenue_new.loc[i] = new_row
        i = i+1
        
#s = np.array(revenue_list_nocomma)

#f = s.astype(float)







tesla_revenue_new.dropna(inplace=True)

tesla_revenue_new = tesla_revenue_new[tesla_revenue_new['Revenue'] != ""]

#tesla_revenue_new


['21454', '16934', '18756', '17719', '13757', '11958', '10389', '10744', '8771', '6036', '5985', '7384', '6303', '6350', '4541', '7226', '6824', '4002', '3409', '3288', '2985', '2790', '2696', '2285', '2298', '1270', '1147', '1214', '937', '955', '940', '957', '852', '769', '621', '615', '431', '405', '562', '306', '50', '27', '30', '39', '58', '58', '49', '36', '31', '28', '21', '', '46', '27']


ValueError: could not convert string to float: 

In [ ]:
print(revenue_list_nocomma)

revenue_list_nocomma[7] = np.float64(revenue_list_nocomma[7])
revenue_list_nocomma[7]

print(revenue_list_nocomma)


In [ ]:
#commented out this code since it did not work

#tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace('$',"")

Execute the following lines to remove an null or empty strings in the Revenue column.


In [76]:
tesla_revenue_new.dropna(inplace=True)

tesla_revenue_new = tesla_revenue_new[tesla_revenue_new['Revenue'] != ""]

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
tesla_revenue_new.tail()

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [ ]:
gameStop = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [ ]:
gme_data = gameStop.history(period = "max")

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [ ]:
gme_data.reset_index(inplace = True)

gme_data.head()

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [ ]:
url1 = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"

html_data_gme = requests.get(url1).text

Parse the html data using `beautiful_soup`.


In [ ]:
soup = BeautifulSoup(html_data_gme, "html.parser")



Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [71]:
tables_gameStop = soup.find_all('table')

tables_gameStop[1]

revenue_data_gameStop = tables_gameStop[1].find_all('tr')

revenue_data_gameStop

data_list_gameStop = []

for row in revenue_data_gameStop:
    row_data = row.find_all('td')
    individual_row_data = [data.text.strip() for data in row_data]
    data_list_gameStop = data_list_gameStop + individual_row_data
    

i = 0
while(i<len(data_list_gameStop)):
      data_list_gameStop[i] = data_list_gameStop[i].replace("$","").replace(",","")
      i = i+1
      
print(data_list_gameStop)

dates_list_gameStop = data_list_gameStop[0:len(data_list_gameStop):2]

dates_list_gameStop

revenue_list_gameStop = data_list_gameStop[1:len(data_list_gameStop):2]

revenue_list_gameStop

revenue_list_gameStop_nocomma = []


gme_revenue = pd.DataFrame(columns = ["Date","Revenue"])

i = 0
while(i<len(dates_list_gameStop)):
    new_row = [dates_list_gameStop[i],revenue_list_gameStop[i]]
    gme_revenue.loc[i] = new_row
    i = i+1
    
gme_revenue

['2022-09-30', '21454', '2022-06-30', '16934', '2022-03-31', '18756', '2021-12-31', '17719', '2021-09-30', '13757', '2021-06-30', '11958', '2021-03-31', '10389', '2020-12-31', '10744', '2020-09-30', '8771', '2020-06-30', '6036', '2020-03-31', '5985', '2019-12-31', '7384', '2019-09-30', '6303', '2019-06-30', '6350', '2019-03-31', '4541', '2018-12-31', '7226', '2018-09-30', '6824', '2018-06-30', '4002', '2018-03-31', '3409', '2017-12-31', '3288', '2017-09-30', '2985', '2017-06-30', '2790', '2017-03-31', '2696', '2016-12-31', '2285', '2016-09-30', '2298', '2016-06-30', '1270', '2016-03-31', '1147', '2015-12-31', '1214', '2015-09-30', '937', '2015-06-30', '955', '2015-03-31', '940', '2014-12-31', '957', '2014-09-30', '852', '2014-06-30', '769', '2014-03-31', '621', '2013-12-31', '615', '2013-09-30', '431', '2013-06-30', '405', '2013-03-31', '562', '2012-12-31', '306', '2012-09-30', '50', '2012-06-30', '27', '2012-03-31', '30', '2011-12-31', '39', '2011-09-30', '58', '2011-06-30', '58', '20

Date Revenue
0   2022-09-30   21454
1   2022-06-30   16934
2   2022-03-31   18756
3   2021-12-31   17719
4   2021-09-30   13757
5   2021-06-30   11958
6   2021-03-31   10389
7   2020-12-31   10744
8   2020-09-30    8771
9   2020-06-30    6036
10  2020-03-31    5985
11  2019-12-31    7384
12  2019-09-30    6303
13  2019-06-30    6350
14  2019-03-31    4541
15  2018-12-31    7226
16  2018-09-30    6824
17  2018-06-30    4002
18  2018-03-31    3409
19  2017-12-31    3288
20  2017-09-30    2985
21  2017-06-30    2790
22  2017-03-31    2696
23  2016-12-31    2285
24  2016-09-30    2298
25  2016-06-30    1270
26  2016-03-31    1147
27  2015-12-31    1214
28  2015-09-30     937
29  2015-06-30     955
30  2015-03-31     940
31  2014-12-31     957
32  2014-09-30     852
33  2014-06-30     769
34  2014-03-31     621
35  2013-12-31     615
36  2013-09-30     431
37  2013-06-30     405
38  2013-03-31     562
39  2012-12-31     306
40  2012-09-30      50
41  2012-06-30      27
42  2012-03-31      30
43  2011-12-31      39
44  2011-09-30      58
45  2011-06-30      58
46  2011-03-31      49
47  2010-12-31      36
48  2010-09-30      31
49  2010-06-30      28
50  2010-03-31      21
51  2009-12-31        
52  2009-09-30      46
53  2009-06-30      27

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
gme_revenue.tail()

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [ ]:
make_graph(tesla_data, tesla_revenue_new, 'Tesla')

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [ ]:
make_graph(gme_data, gme_revenue, 'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
